fix: Reduce VM recovery times

jcanocan opened the pull request:
**What this PR does / why we need it**:
It aims to reduce the VM recovery time when an operator such as [NodeHealthCheck](https://github.com/medik8s/node-healthcheck-operator/) is used to remediate unhealthy nodes, i.e. nodes no longer responsive.

In such scenario, the launcher pod is deleted, and the VMi needs to be rescheduled if the spec `runStrategy` is configured as `Always` or `RerunOnFailure`. The VM's Phase is `Running` but VM has the following condition:

```
Status:
  Conditions:
    Last Probe Time:       2023-12-18T16:31:31Z
    Last Transition Time:  2023-12-18T16:31:31Z
    Message:               virt-launcher pod has not yet been scheduled
    Reason:                PodNotExists
```

Eventually, the VMi is rescheduled and the virt-launcher pod recreated.  However, it takes over than 100 seconds in best case. In the following figure, the total VM recovery time break down can be seen where:

- Vm_not_ready: Time for the VMi to be READY=False.
- Launcher_terminating: Time until launcher pod changes state to terminating.
- Launcher_deleted: Time until the launcher pod is deleted from the cluster.
- Launcher_respawned: Time until the launcher pod is respawned (respanwed: in any state pending, running, …).
- Vm_ready: Time for the VMi to be READY=True.

Please note that those measures are taken from the previous step until the step is finished. Regarding the vm_not_ready measurement, this time starts from the point where kubelet is stopped (this is the way to force a node to be unhealthy) until this VMi state is reached.
For instance, launcher_terminating measurement starts when vm_not_ready, i.e.  the VMi READY field is False, until the launcher pod updates its status to terminating.

![kubevirtci-vm-break-down](https://github.com/kubevirt/kubevirt/assets/110104425/82f4c813-c7de-40f0-bb48-0cdab9d450ce)

As can be seen, the vm_launcher_respawned speed-up is approximately 100x.

The issue is: if a VMi in Phase `Running` on a given node, and the node becomes unhealthy, VMis are not marked as `Failed`. This is due to the fact that the main responsible for this task is the virt-handler pod, which own the VMi. However, the virt-handler pod will be unresponsive, unable to be redeployed and unable to mark the VMi as `Failed`.  The secondary responsible for this action is the node controller. Nevertheless, the node controller requires that heartbeats from the virt-handler have not been received for a certain period of time (5 minutes).

Therefore, in order to transition the VMi to a FAILED state, it is required to wait until:

1. The node to get healthy again, the virt-handler to be recreated, the virt-handler to detect that VMi is on `Running` phase without a launcher-pod, and to mark the VMi as `Failed`. 

https://github.com/kubevirt/kubevirt/blob/4d095205df627bff1fe6ae10495e281a1475edc6/pkg/virt-handler/vm.go#L3280-L3283

2. The node controller to detect that the virt-handler has been unresponsive for a given amount of time (5 minutes by default), to detect VMis without virt-launcher pod and to mark the VMis as `Failed`.

https://github.com/kubevirt/kubevirt/blob/d4e5e11c84d0630c1c688b1bfffe7a0133f9675d/pkg/virt-controller/watch/node.go#L462-L477

It reduces the virt-launcher pod respawn time by short-circuiting the feedback loop that the node controller would already handle and the virt-handler recreation. Therefore, the virt-controller will mark the VMi as `Failed` and it will force the VMi to be rescheduled on another node faster.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes # [CNV-36137](https://issues.redhat.com/browse/CNV-36137)

**Special notes for your reviewer**:

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Reduced VM rescheduling time on node failure
```

aburdenthehand commented:
/cc @enp0s3

xpivarc commented:
/cc

xpivarc commented:
@jcanocan Do we have a function test simulating this? It seems to me as easily reproducible in the test.

jcanocan commented:
> @jcanocan Do we have a function test simulating this? It seems to me as easily reproducible in the test.

I didn't have the chance to try out. However, the test should work without any external dependencies such as NHC, right?

xpivarc commented:
> > @jcanocan Do we have a function test simulating this? It seems to me as easily reproducible in the test.
> 
> I didn't have the chance to try out. However, the test should work without any external dependencies such as NHC, right?

Right.

jcanocan commented:
> > > @jcanocan Do we have a function test simulating this? It seems to me as easily reproducible in the test.
> > 
> > 
> > I didn't have the chance to try out. However, the test should work without any external dependencies such as NHC, right?
> 
> Right.

Ok, I will try my best. If you have an idea or an example, that would be helpful. Right now, I'm trying to solve the failures.

jcanocan commented:
v1:
- Fixed unit test to adjust new status updates since if the pod is not present, it will be created in a single sync call. This causes new updates and patches on the pods and VMis no previously considered.

xpivarc commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -1139,24 +1156,24 @@ func (c *VMIController) hotplugPodsReady(vmi *virtv1.VirtualMachineInstance, vir
 }
 
-func (c *VMIController) sync(vmi *virtv1.VirtualMachineInstance, pod *k8sv1.Pod, dataVolumes []*cdiv1.DataVolume) syncError {
+func (c *VMIController) sync(vmi *virtv1.VirtualMachineInstance, pod *k8sv1.Pod, dataVolumes []*cdiv1.DataVolume) (*k8sv1.Pod, syncError) {
```
Why do you need to return `Pod`?

jcanocan replied:
Because AFAIK, the intention in [line](https://github.com/kubevirt/kubevirt/pull/10961/files#diff-16487fff595728eb5eeec01225a2203756ecad25d90e90d5feb13a9ff08df7ffR1232) 


```go
pod, err := c.clientset.CoreV1().Pods(vmi.GetNamespace()).Create(context.Background(), templatePod, v1.CreateOptions{})
		//this will fail since pod is nil
		if k8serrors.IsForbidden(err) && strings.Contains(err.Error(), "violates PodSecurity") {
			psaErr := fmt.Errorf("failed to create pod for vmi %s/%s, it needs a privileged namespace to run: %w", vmi.GetNamespace(), vmi.GetName(), err)
			c.recorder.Eventf(vmi, k8sv1.EventTypeWarning, FailedCreatePodReason, failedToRenderLaunchManifestErrFormat, psaErr)
			return &syncErrorImpl{psaErr, FailedCreatePodReason}
			return pod, &syncErrorImpl{psaErr, FailedCreatePodReason}
		}
```

is if the pod is not found, we need to create it, assign it to `pod` variable and reflect the change outside the `sync` function in order to update the vmi state later on. So, if I'm right, please correct me if I'm wrong, as it is the `pod` variable change won't be seen outside the sync function.

xpivarc replied:
It is not required because the informer will enqueue the VMI once it observes that we have created the Pod.

jcanocan replied:
Right, but it takes time, which is what I'm trying to reduce here. For putting in context, as it is, the launcher_respawn time is nearly 1 second, and without retuning the pod the launcher_respawn time is 20 seconds. I didn't have the chance to try out on much more complex scenarios. However, I expect this time to be much higher.

xpivarc replied:
This should not be a problem. An informer should observe changes and it should trigger the next resync fast enough for us to observe. Could we retest and investigate this if you observe this again?

0xFelix replied:
Currently there is the chance for a gap between creation of the Pod and reenqueing it for another run to update the VM's status. The informer should be pretty fast, but there is a slight chance that the VM's status and the reality (Pod already created) become out of sync. IMO this is a neat and simple way to create the Pod and update the VM's status in one go.

jcanocan replied:
I've carried out a much more detailed experimentation. The result is this change does not have a meaningful impact on the final recovery time. Therefore, you are right, and this change has been reverted. Nevertheless, IMHO, we should consider this change in the future.

xpivarc commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -870,6 +871,22 @@ func prepareVMIPatch(oldVMI, newVMI *virtv1.VirtualMachineInstance) ([]byte, err
 		log.Log.V(3).Object(oldVMI).Infof("Patching VMI activePods")
 	}
 
+	if !equality.Semantic.DeepEqual(newVMI.Status.Phase, oldVMI.Status.Phase) {
```
This should be cover by
```
// If we detect a change on the vmi we update the vmi
	vmiChanged := !equality.Semantic.DeepEqual(vmi.Status, vmiCopy.Status) || !equality.Semantic.DeepEqual(vmi.Finalizers, vmiCopy.Finalizers) || !equality.Semantic.DeepEqual(vmi.Annotations, vmiCopy.Annotations) || !equality.Semantic.DeepEqual(vmi.Labels, vmiCopy.Labels)
	if vmiChanged {
		key := controller.VirtualMachineInstanceKey(vmi)
		c.vmiExpectations.SetExpectations(key, 1, 0)
		_, err := c.clientset.VirtualMachineInstance(vmi.Namespace).Update(context.Background(), vmiCopy)
		if err != nil {
			c.vmiExpectations.LowerExpectations(key, 1, 0)
			return err
		}
	}

	return nil
```

jcanocan replied:
As far as I was able to test, it is not. You're completely right that this code should update the VMi Phase. But for some reason, when the node gets down, it is not updating the VMi Phase.

xpivarc replied:
So this sounds like a bug to me. Could you have a look at this?

jcanocan replied:
I've taken a look at it and the update part is not executed until the VMi is already in `Failed` state. 

Looking at this [line](https://github.com/jcanocan/kubevirt/blob/56012c131325f2bb24714bf69ea40beb35624418/pkg/virt-controller/watch/vmi.go#L752) :

```go
// VMI is owned by virt-handler, so patch instead of update
if vmi.IsRunning() || vmi.IsScheduled() {
		patchBytes, err := prepareVMIPatch(vmi, vmiCopy)
		if err != nil {
			return fmt.Errorf("error preparing VMI patch: %v", err)
		}

		if len(patchBytes) > 0 {

			_, err = c.clientset.VirtualMachineInstance(vmi.Namespace).Patch(context.Background(), vmi.Name, types.JSONPatchType, []byte(patchBytes), &v1.PatchOptions{})
			// We could not retry if the "test" fails but we have no sane way to detect that right now: https://github.com/kubernetes/kubernetes/issues/68202 for details
			// So just retry like with any other errors
			if err != nil {
				return fmt.Errorf("patching of vmi conditions and activePods failed: %v", err)
			}
		}

		return nil
	}

```
it makes sense since the VMi is on `Running` state. Therefore, it will execute the patch section of code and `return nil`, so the update is not executed until the VMi is on a `Phase` different from `Running` or ` Scheduled`. IMHO, this is right, and we should keep the `Phase` update in the patch section.

If we don't patch the `Phase` in this point, the launcher_respawn time is the same as the baseline case (up to 100s).

0xFelix replied:
As long as the VMI is considered running/scheduled, `virt-controller` will only patch it. As @jcanocan pointed out, therefore Update is never called. See L767.

xpivarc replied:
You are both correct, sorry for the confusion.

Could we change this slightly:

```
if newVMI.Status.Phase == v1.Failed &&  oldVMI.Status.Phase == v1.Running  && pod == nil{
    patchOps = append(patchOps,  `{ "op": "test", "path": "/status/phase", "value": "Running" }`)
    patchOps = append(patchOps, `{ "op": "replace", "path": "/status/phase", "value": "Failed" }`)
}
```
Note: we need to pass the pod into the function. 
Note 2: no need for the `Marshal`

The reason behind this is to limit when we make this patch as much as we can. Right now we have a clear separation of when the handler or controller acts and I want to make sure this is seen as an exception. Also comment would be really good to have here.

jcanocan replied:
Done.

xpivarc commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -708,7 +708,8 @@ func (c *VMIController) updateStatus(vmi *virtv1.VirtualMachineInstance, pod *k8
 
 	case vmi.IsRunning():
-		if !vmiPodExists {
+		if !vmiPodExists || isPodDownOrGoingDown(pod) {
```
This is wrong. In a normal situation at this point, the owner of the VMI is the handler and only the handler should move the VMI to failed. Here we need to express, "Pod is down, the handler is down, and let's move the VMI to Failed".

If we do not do this then 99% of the time the controller and handler would be racing.

xpivarc replied:
I think changing this to `podIsDown` might do the trick. What is left to do is to see if migration doesn't influence this.

jcanocan replied:
Done. Please, check if I got it right. Thanks!

xpivarc commented on pkg/virt-controller/watch/vmi_test.go:
```diff
@@ -189,6 +189,12 @@ var _ = Describe("VirtualMachineInstance watcher", func() {
 		})
 	}
 
+	ignorePodPatches := func() {
```
This is a hack and should not be required.

jcanocan replied:
If we don't do something similar, the test suite will fail because there are unexpected patches to pods. This is due to the fact that if a pod is not found, now it will create it in line:

```go
pod, err := c.clientset.CoreV1().Pods(vmi.GetNamespace()).Create(context.Background(), templatePod, v1.CreateOptions{})
		//this will fail since pod is nil
		if k8serrors.IsForbidden(err) && strings.Contains(err.Error(), "violates PodSecurity") {
			psaErr := fmt.Errorf("failed to create pod for vmi %s/%s, it needs a privileged namespace to run: %w", vmi.GetNamespace(), vmi.GetName(), err)
			c.recorder.Eventf(vmi, k8sv1.EventTypeWarning, FailedCreatePodReason, failedToRenderLaunchManifestErrFormat, psaErr)
			return &syncErrorImpl{psaErr, FailedCreatePodReason}
			return pod, &syncErrorImpl{psaErr, FailedCreatePodReason}
		}
```
and later on, in `syncPausedConditionToPod` function, we will patch the pod. Therefore, something similar is required, otherwise, the test will fail. Do you know another way around, less hacky?

xpivarc replied:
I don't understand how this fails. Is the Pod not there?

jcanocan replied:
No, in these tests where I've added the function `ignorePodPatched` the pod is not present before starting the `controller.Execute()` function. Therefore, it needs to be created. The function [sync](https://github.com/kubevirt/kubevirt/pull/10961/files#diff-16487fff595728eb5eeec01225a2203756ecad25d90e90d5feb13a9ff08df7ffR1203) creates it. Before the change, due to the fact that the pod is created, but it is not returned (therefore after `sync` [function](https://github.com/jcanocan/kubevirt/blob/56012c131325f2bb24714bf69ea40beb35624418/pkg/virt-controller/watch/vmi.go#L391) 
the variable `pod` is still `pod = nil`) this made that `updateStatus` [function](https://github.com/jcanocan/kubevirt/blob/56012c131325f2bb24714bf69ea40beb35624418/pkg/virt-controller/watch/vmi.go#L525) does not execute function `syncPausedConditionToPod` [function](https://github.com/jcanocan/kubevirt/blob/56012c131325f2bb24714bf69ea40beb35624418/pkg/virt-controller/watch/vmi.go#L1030). Now, function `syncPausedConditionToPod` is executed (because the pod != nil) and it triggers a patch in the pod. Therefore, it fails if `ignorePodPatches` is not added because the test cases don't expect pod patches.

jcanocan replied:
This function has been removed.

jcanocan commented:
v2:
- Added a check to only mark VMis as failed when the virt-handler pod is down or gone.

With this additional check, the launcher_pod are not respawned as fast as the original approach. However, it does not break current functional test and the VMi behavior are identical. Check:

![Figure_1](https://github.com/kubevirt/kubevirt/assets/110104425/89049350-8292-4e37-a306-cfad8df4ebd0)

The main source of delay has been spotted when this error appears on virt-controller logs:

```
{"component":"virt-controller","level":"info","msg":"reenqueuing VirtualMachineInstance default/fedora-test-snr","pos":"vmi.go:325","reason":"patching of vmi conditions and activePods failed: Internal error occurred: failed calling webhook \"virtualmachineinstances-update-validator.kubevirt.io\": failed to call webhook: Post \"https://virt-api.kubevirt.svc:443/virtualmachineinstances-validate-update?timeout=10s\": context deadline exceeded","timestamp":"2024-01-05T13:53:44.490347Z"}
I0105 13:53:44.490389       1 trace.go:236] Trace[945221119]: "virt-controller VMI workqueue" Workqueue Key:default/fedora-test-snr (05-Jan-2024 13:53:34.481) (total time: 10008ms):
Trace[945221119]: ---"updateStatus" VMI Name:fedora-test-snr 10008ms (13:53:44.490)
Trace[945221119]: [10.008402325s] [10.008402325s] END
```

jcanocan commented:
/cc @0xFelix

jcanocan commented:
/retest-required
The failing test works locally, giving it another shoot.

jcanocan commented:
The failing test works locally, giving it another shoot.
/retest-required

jcanocan commented:
/retest-required

0xFelix commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -27,6 +27,9 @@ import (
 	"strings"
 	"time"
 
+	"k8s.io/apimachinery/pkg/fields"
```
Move imports to block starting at L39

jcanocan replied:
Done.

0xFelix commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -709,6 +712,16 @@ func (c *VMIController) updateStatus(vmi *virtv1.VirtualMachineInstance, pod *k8
+			virtHandlerAlive, err := isVirtHandlerAlive(c.clientset, vmi.Status.NodeName)
+
+			if err != nil {
+				log.Log.Errorf("virt-handler pod is not responsible %v", err)
```
That error message sounds weird. IMO it should read `virt-handler pod is not alive ...`. Also it could be a warning instead of an error.

jcanocan replied:
I've dropped this message.

0xFelix commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -778,6 +791,29 @@ func (c *VMIController) updateStatus(vmi *virtv1.VirtualMachineInstance, pod *k8
 	return nil
 }
 
+func isVirtHandlerAlive(clientset kubecli.KubevirtClient, nodeName string) (bool, error) {
```
I'd change this function so it returns `err != nil` only in cases where an actual error happened. Coming to the conclusion that no virt handler is active is not an error?

jcanocan replied:
Done.

davidvossel commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -778,6 +791,29 @@ func (c *VMIController) updateStatus(vmi *virtv1.VirtualMachineInstance, pod *k8
+		return false, fmt.Errorf("virt-handler pod not found")
+	}
+
+	for _, pod := range list.Items {
+		if podIsDown(&pod) || isHandlerContainerFailed(&pod) || pod.DeletionTimestamp != nil {
+			return false, fmt.Errorf("virt-handler is going down, node %s pod %s", nodeName, pod.Name)
+		}
+	}
```
It might be better to reverse this function and look for a running virt-handler in order to determine if virt-handler is alive rather than looking for a dead virt-handler.

for example, consider updates here. It's possible during a rolling update that we'll have a virt-handler marked for deletion `DeletionTimestamp != nil` and a new virt-handler pod that is active.

jcanocan replied:
You are absolutely right. Thanks! Done.

davidvossel commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -778,6 +791,29 @@ func (c *VMIController) updateStatus(vmi *virtv1.VirtualMachineInstance, pod *k8
+		LabelSelector: handlerLabelSelector.String(),
+	})
+
+	if err != nil {
+		return false, fmt.Errorf("failed to get virt-handler pods %v", err)
+	}
```
A failure to get virt-handler pods should not result in virt-controller marking vmi's as failed. that's an api failure that doesn't reflect whether there's a virt-handler online or not.

jcanocan replied:
Done.

davidvossel commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -709,6 +712,16 @@ func (c *VMIController) updateStatus(vmi *virtv1.VirtualMachineInstance, pod *k8
+				log.Log.Errorf("virt-handler pod is not responsible %v", err)
+			}
+
+			if !virtHandlerAlive && (!vmiPodExists || isPodDownOrGoingDown(pod)) {
+				vmiCopy.Status.Phase = virtv1.Failed
+			}
```
the`&& (!vmiPodExists || isPodDownOrGoingDown(pod)` part of this isn't necessary or even accurate. The pod is most likely `nil` here which would cause `isPodDownOrGoingDown` to segfault.

you're already in an if clause that's checking if `!vmiPodExists` and there shouldn't be a reason to look at `pod`

jcanocan replied:
Done.

davidvossel reviewed (commented):
Without your patch, what eventually transitions the VMI to `phase=Failed`.  Are we dependent on virt-handler eventually coming back online on the same node that the VMI is scheduled to?

It seems like without your patch here, we could be in a situation where if a node is abruptly killed and never comes back online, that vmi's with phase=Running might be stuck indefinitely in that phase.

xpivarc commented on pkg/virt-controller/watch/vmi_test.go:
```diff
@@ -1769,6 +1769,17 @@ var _ = Describe("VirtualMachineInstance watcher", func() {
 
 			addVirtualMachine(vmi)
 
+			if phase == virtv1.Running {
```
You want to replace this by 
```
   pod := <some minimalistic Pod spec>
   kubeClient ... Create(pod)
```
and then you don't need a reactor. Think about kubeClient as an API server and you want to populate it with the `virt-handler` Pod.

jcanocan replied:
IIUC, the idea is to create a `virt-handler` pod in this test so we can avoid the reactor. However, this is not possible due to the fact that in this test the `virt-launcher` is not present, and it will call the new `isVirtHandlerAlive` function, which needs to list the virt-handler pods. Therefore, the reactor is still needed.
Another possibility is to create `virt-launcher` to avoid the reactor (the `virt-launcher` will be present and the `isVirtHandlerAlive` won't be called). However, I'm unsure if the test scenario wants to have the `virt-launcher` pod. WDYT?

xpivarc reviewed (commented):
What is the current improvement?

xpivarc commented:
@davidvossel @vladikr 
Initially, I thought it could be better to check for the handler as well but now when I think about it, this could be as simple as 
```
if !vmiPodExists {
   vmi.Status.Phase = v1.Failed
```
as I don't see any negatives of doing it. WDYT?

0xFelix commented:
> Without your patch, what eventually transitions the VMI to `phase=Failed`. Are we dependent on virt-handler eventually coming back online on the same node that the VMI is scheduled to?
> 
> It seems like without your patch here, we could be in a situation where if a node is abruptly killed and never comes back online, that vmi's with phase=Running might be stuck indefinitely in that phase.

From what I've understood, right now we are dependent on virt-controller eventually picking up that the virt-handler on a node died (no heartbeat). In @jcanocan's test scenario he kills a node, that comes back after a while. The node coming back online triggers the node controller in virt-controller, which eventually marks the VMI as Failed. Until then the VMI remains in phase Running, even though the virt-launcher Pod does not exist. This PR tries to improve that behavior, by marking the VMI as Failed, when virt-controller sees, that it is running but there is no Pod. (basically what @xpivarc suggested in his last comment)

See https://github.com/kubevirt/kubevirt/blob/d4e5e11c84d0630c1c688b1bfffe7a0133f9675d/pkg/virt-controller/watch/node.go#L186

jcanocan commented:
> @davidvossel @vladikr Initially, I thought it could be better to check for the handler as well but now when I think about it, this could be as simple as
> 
> ```
> if !vmiPodExists {
>    vmi.Status.Phase = v1.Failed
> ```
> 
> as I don't see any negatives of doing it. WDYT?

I've spotted some issues with VMReplicaSets.

davidvossel commented:
> Initially, I thought it could be better to check for the handler as well but now when I think about it, this could be as simple as
> if !vmiPodExists {
>   vmi.Status.Phase = v1.Failed
>as I don't see any negatives of doing it. WDYT?

looking at this closer. I think you are correct. And here's why

1. This is same logic that the node controller is doing today when the virt-handler heartbeat times out https://github.com/kubevirt/kubevirt/blob/4d095205df627bff1fe6ae10495e281a1475edc6/pkg/virt-controller/watch/node.go#L248 
2. Virt-handler already knows how to handle cleaning up local cache even after the vmi is removed from the api server.  the domain informer has a local cache it uses to know if a vmi used to exist https://github.com/kubevirt/kubevirt/blob/4d095205df627bff1fe6ae10495e281a1475edc6/pkg/virt-handler/vm.go#L1624 

Essentially what we're doing by moving a `vmi.Status.Phase == Running` to `vmi.Status.Phase == Failed` is shortening the feedback loop that the node controller would already handle.

I think we can take this a step further even and move the VMI to failed if the vmi is Scheduled or Running when no vmi pod exists.

jcanocan commented:
v3:
- Added a functional test.

xpivarc commented:
> > Initially, I thought it could be better to check for the handler as well but now when I think about it, this could be as simple as
> > if !vmiPodExists {
> > vmi.Status.Phase = v1.Failed
> > as I don't see any negatives of doing it. WDYT?
> 
> looking at this closer. I think you are correct. And here's why
> 
>     1. This is same logic that the node controller is doing today when the virt-handler heartbeat times out https://github.com/kubevirt/kubevirt/blob/4d095205df627bff1fe6ae10495e281a1475edc6/pkg/virt-controller/watch/node.go#L248

Exactly, this is just short-circuiting it in case a Pod is explicitly deleted by humans or automation.

> 
>     2. Virt-handler already knows how to handle cleaning up local cache even after the vmi is removed from the api server.  the domain informer has a local cache it uses to know if a vmi used to exist https://github.com/kubevirt/kubevirt/blob/4d095205df627bff1fe6ae10495e281a1475edc6/pkg/virt-handler/vm.go#L1624
> 

+1 I believe this is covered by tests as well.

> 
> Essentially what we're doing by moving a `vmi.Status.Phase == Running` to `vmi.Status.Phase == Failed` is shortening the feedback loop that the node controller would already handle.
>

+1 

> I think we can take this a step further even and move the VMI to failed if the vmi is Scheduled or Running when no vmi pod exists.

@jcanocan feel free to drop the check for virt-handler.

jcanocan commented:
v4:
- Dropped the virt-handler is alive check.
- VMi will be marked as `Failed` also if the VMi is `Scheduled` but the virt-launcher pod is not found.
- Updated commit description.

jcanocan commented:
v4:
- Dropped the virt-handler is alive check.
- VMi will be marked as `Failed` also if the VMi is `Scheduled` but the virt-launcher pod is not found.
- Updated commit description.

jcanocan commented:
> > > Initially, I thought it could be better to check for the handler as well but now when I think about it, this could be as simple as
> > > if !vmiPodExists {
> > > vmi.Status.Phase = v1.Failed
> > > as I don't see any negatives of doing it. WDYT?
> > 
> > 
> > looking at this closer. I think you are correct. And here's why
> > ```
> > 1. This is same logic that the node controller is doing today when the virt-handler heartbeat times out https://github.com/kubevirt/kubevirt/blob/4d095205df627bff1fe6ae10495e281a1475edc6/pkg/virt-controller/watch/node.go#L248
> > ```
> 
> Exactly, this is just short-circuiting it in case a Pod is explicitly deleted by humans or automation.
> 
> > ```
> > 2. Virt-handler already knows how to handle cleaning up local cache even after the vmi is removed from the api server.  the domain informer has a local cache it uses to know if a vmi used to exist https://github.com/kubevirt/kubevirt/blob/4d095205df627bff1fe6ae10495e281a1475edc6/pkg/virt-handler/vm.go#L1624
> > ```
> 
> +1 I believe this is covered by tests as well.
> 
> > Essentially what we're doing by moving a `vmi.Status.Phase == Running` to `vmi.Status.Phase == Failed` is shortening the feedback loop that the node controller would already handle.
> 
> +1
> 
> > I think we can take this a step further even and move the VMI to failed if the vmi is Scheduled or Running when no vmi pod exists.
> 
> @jcanocan feel free to drop the check for virt-handler.

Thanks for all the feedback and insights provided. This is very helpful. I've updated the PR with the new suggestions. I'm not sure if you want me to mark the VMI as failed in this PR or on a follow-up. In any case, I will take the chance to test it out.

jcanocan commented:
/retest-required

jcanocan commented:
/retest-required

0xFelix commented on tests/vm_test.go:
```diff
@@ -1921,6 +1923,92 @@ status:
+			go func() {
+				defer wg.Done()
+				By("The VMi should transition to phase Failed")
+				Eventually(vmiWatch.ResultChan(), 60*time.Second).Should(Receive(WithTransform(func(e watch.Event) v1.VirtualMachineInstanceStatus {
```
Is it necessary to do it with a watch? Wouldn't it be enough to poll with Eventually, or could you miss the phase change then?

jcanocan commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -870,6 +873,16 @@ func prepareVMIPatch(oldVMI, newVMI *virtv1.VirtualMachineInstance) ([]byte, err
 	}
 
+	// This patch represents an exception on which the VMI is handled by the virt-handler, but it is on an unhealthy node,
+	//i.e. the virt-handler is gone or terminating.
```
Done.

jcanocan commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -870,6 +873,16 @@ func prepareVMIPatch(oldVMI, newVMI *virtv1.VirtualMachineInstance) ([]byte, err
 
+	// This patch represents an exception on which the VMI is handled by the virt-handler, but it is on an unhealthy node,
+	//i.e. the virt-handler is gone or terminating.
+	if (oldVMI.Status.Phase == virtv1.Running || oldVMI.Status.Phase == virtv1.Scheduled) && newVMI.Status.Phase == virtv1.Failed && pod == nil {
```
Done.

jcanocan commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -1108,6 +1121,15 @@ func isComputeContainerFailed(pod *k8sv1.Pod) bool {
 	return false
 }
 
+func isHandlerContainerFailed(pod *k8sv1.Pod) bool {
```
Nice catch. It's not. Removed.

jcanocan commented on tests/libnode/node.go:
```diff
@@ -256,14 +268,30 @@ func Taint(nodeName string, key string, effect k8sv1.TaintEffect) {
+}
 
-	patchBytes, err := strategicpatch.CreateTwoWayMergePatch(old, newJson, node)
+func UnTaint(nodeName string, key string) {
```
Done.

jcanocan commented on tests/vm_test.go:
```diff
@@ -1921,6 +1923,92 @@ status:
+			node.Spec.Unschedulable = true
+			node, err = kubevirt.Client().CoreV1().Nodes().Update(context.Background(), node, metav1.UpdateOptions{})
+			Expect(err).ToNot(HaveOccurred())
+			Expect(node.Spec.Unschedulable).To(BeTrue())
```
I don't think so. Removed.

jcanocan commented on tests/vm_test.go:
```diff
@@ -1921,6 +1923,92 @@ status:
+
+			By("VMi should be NotReady")
+
+			Eventually(ThisVMI(vmi), 300*time.Second, 2*time.Second).Should(HaveConditionFalse(v1.VirtualMachineInstanceReady))
```
Reduced to 1 min.

jcanocan commented on tests/vm_test.go:
```diff
@@ -1921,6 +1923,92 @@ status:
+			vm, err := kubevirt.Client().VirtualMachine(vm.Namespace).Create(context.Background(), vm)
+			Expect(err).ToNot(HaveOccurred())
+			Eventually(ThisVM(vm), 360*time.Second, 1*time.Second).Should(beReady())
+			vmi = libwait.WaitForSuccessfulVMIStart(vmi)
```
You are right. Removed.

jcanocan commented on tests/libnode/node.go:
```diff
@@ -243,6 +243,18 @@ func RemoveAnnotationFromNode(nodeName string, key string) *k8sv1.Node {
 }
 
+func patchNode(virtCli kubecli.KubevirtClient, old []byte, newNode *k8sv1.Node) {
+
```
Done.

jcanocan commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -1178,6 +1200,7 @@ func (c *VMIController) sync(vmi *virtv1.VirtualMachineInstance, pod *k8sv1.Pod,
 	}
 
 	if !podExists(pod) {
+
```
Done.

jcanocan commented:
v5:
- Fixed a couple of nits and leftovers from previous versions.
- Updated `case vmi.IsScheduled():`. Now, the `break` statement is outside the `if !vmiPodExists`.
- Reduced VM `NotReady` timeout and no required checks on functional test.
- Updated patch comment to better reflect why it is required.

jcanocan commented:
/retest-required

jcanocan commented:
/retest pull-kubevirt-check-tests-for-flakes

jcanocan commented:
/test pull-kubevirt-check-tests-for-flakes

jcanocan commented:
/test pull-kubevirt-check-tests-for-flakes

jcanocan commented:
/test pull-kubevirt-check-tests-for-flakes

xpivarc commented:
/hold

> /test pull-kubevirt-check-tests-for-flakes

This is a positive finding, please fix the test.

jcanocan commented:
> > /hold
> > > /test pull-kubevirt-check-tests-for-flakes
> > 
> > 
> > This is a positive finding, please fix the test.
> 
> Done. Let's see how it goes now.

@xpivarc looks like it is much better now!

0xFelix commented on tests/vm_test.go:
```diff
@@ -1921,6 +1922,91 @@ status:
+			node.Spec.Unschedulable = false
+			node, err = kubevirt.Client().CoreV1().Nodes().Update(context.Background(), node, metav1.UpdateOptions{})
+			Expect(err).ToNot(HaveOccurred())
+			Expect(node.Spec.Unschedulable).To(BeFalse())
```
Not absolutely necessary to check again, if the Update succeeded.

jcanocan replied:
Done.

0xFelix commented on tests/vm_test.go:
```diff
@@ -1921,6 +1922,91 @@ status:
+		})
+		It("[Serial] the VMs running in that node should be respawned", func() {
+			var pod *k8sv1.Pod
+			var wg sync.WaitGroup
```
Move defines to L1975 and L1982?

0xFelix commented on tests/vm_test.go:
```diff
@@ -1921,6 +1922,91 @@ status:
+			By("Starting VM")
+
+			vmi := libvmi.NewAlpine()
+			vmi.Namespace = testsuite.GetTestNamespace(vmi)
```
Not necessary, you can inline `libvmi.NewAlpine` into `createRunningVM` and drop this line.

jcanocan replied:
Done

0xFelix commented on tests/vm_test.go:
```diff
@@ -1921,6 +1922,91 @@ status:
+
+			vmi := libvmi.NewAlpine()
+			vmi.Namespace = testsuite.GetTestNamespace(vmi)
+			vm := createRunningVM(virtClient, vmi)
+
+			vm = startVM(virtClient, vm)
```
Use `createVM` instead? `createRunningVM` sets the running field, while startVM uses the run strategy.

0xFelix replied:
Also you could inline createVM into startVM.

jcanocan replied:
Done

0xFelix commented on tests/vm_test.go:
```diff
@@ -1921,6 +1922,91 @@ status:
+			Expect(err).ToNot(HaveOccurred())
+			Expect(node.Spec.Unschedulable).To(BeFalse())
+
+			libnode.Untaint(nodeName, "node.kubernetes.io/unschedulable")
```
Make this label a const?

jcanocan replied:
Done

0xFelix commented on tests/vm_test.go:
```diff
@@ -1921,6 +1922,91 @@ status:
+			node, err = kubevirt.Client().CoreV1().Nodes().Update(context.Background(), node, metav1.UpdateOptions{})
+			Expect(err).ToNot(HaveOccurred())
+
+			By("VMi should be NotReady")
```
nit: VMI

jcanocan replied:
Done

0xFelix commented on tests/vm_test.go:
```diff
@@ -1921,6 +1922,91 @@ status:
+				By("The VMi should transition to phase Failed")
+				Eventually(vmiWatch.ResultChan(), 60*time.Second).Should(Receive(WithTransform(func(e watch.Event) v1.VirtualMachineInstanceStatus {
+					vmi, ok := e.Object.(*v1.VirtualMachineInstance)
+					Expect(ok).To(BeTrue())
```
`Expect` used like this in Eventually can make the whole test fail.

0xFelix commented on tests/vm_test.go:
```diff
@@ -1921,6 +1922,91 @@ status:
+					vmi, ok := e.Object.(*v1.VirtualMachineInstance)
+					Expect(ok).To(BeTrue())
+					return vmi.Status
+				}, gstruct.MatchFields(gstruct.IgnoreExtras, gstruct.Fields{
```
Return the Phase instead and match on that to keep it simple?

jcanocan replied:
Dropped.

0xFelix commented on tests/vm_test.go:
```diff
@@ -1921,6 +1922,91 @@ status:
+			wg.Add(1)
+
+			go func() {
+				defer wg.Done()
```
Is this defer called, if the test fails?

jcanocan replied:
Dropped.

0xFelix commented on tests/vm_test.go:
```diff
@@ -1921,6 +1922,91 @@ status:
+				return pod.DeletionTimestamp != nil
+			}, 30*time.Second, 2*time.Second).Should(BeTrue())
+
+			vmiWatch, err := kubevirt.Client().VirtualMachineInstance(testsuite.GetTestNamespace(vmi)).Watch(context.Background(), metav1.ListOptions{})
```
Move definition of vmiWatch to above where it is used? Inside the goroutine.

jcanocan replied:
Dropped.

0xFelix commented on tests/vm_test.go:
```diff
@@ -1921,6 +1922,91 @@ status:
+				metav1.DeleteOptions{GracePeriodSeconds: pointer.Int64(0)})
+			Expect(err).ToNot(HaveOccurred())
+
+			wg.Wait()
```
Can this block the test suite?

jcanocan replied:
Dropped.

0xFelix commented on tests/vm_test.go:
```diff
@@ -1921,6 +1922,91 @@ status:
+			vmi, err = virtClient.VirtualMachineInstance(vm.Namespace).Get(context.Background(), vm.Name, &k8smetav1.GetOptions{})
+			Expect(err).ToNot(HaveOccurred())
+
+			By("Taint the node, it simulates that something went wrong in the cluster and workloads needs to be removed")
+
+			nodeName = vmi.Status.NodeName
+			libnode.Taint(nodeName, "node.kubernetes.io/unschedulable", k8sv1.TaintEffectNoExecute)
+
+			node, err := kubevirt.Client().CoreV1().Nodes().Get(context.Background(), nodeName, metav1.GetOptions{})
+			Expect(err).ToNot(HaveOccurred())
+
+			By("Disable scheduling in the node")
+
+			node.Spec.Unschedulable = true
+			node, err = kubevirt.Client().CoreV1().Nodes().Update(context.Background(), node, metav1.UpdateOptions{})
+			Expect(err).ToNot(HaveOccurred())
```
Move this into a helper func?

jcanocan replied:
Done.

xpivarc reviewed (commented):
Minor changes to the tests and I think we are done.

jcanocan commented on pkg/virt-controller/watch/vmi_test.go:
```diff
@@ -1769,6 +1769,9 @@ var _ = Describe("VirtualMachineInstance watcher", func() {
 
 			addVirtualMachine(vmi)
 
+			patch := fmt.Sprintf(`[{ "op": "test", "path": "/status/phase", "value": "%s" }, { "op": "replace", "path": "/status/phase", "value": "Failed" }]`, phase)
```
Adjusted the name. 

In regard to add a new test, Isn't that case already covered in test `should do nothing if pod is handed to virt-handler`?

https://github.com/jcanocan/kubevirt/blob/ff9afcabb0790660a66579c0e7cf33c0424ba659/pkg/virt-controller/watch/vmi_test.go#L1603

jcanocan commented on tests/vm_test.go:
```diff
@@ -1921,6 +1922,91 @@ status:
 		})
 	})
+
+	Context("when node becomes unhealthy", Serial, func() {
```
Done. I've left `[Serial]` both in the `Context` and `It` descriptions, please confirm if this is correct. Should I also add the decorator `Serial` to the `It` statement?

jcanocan commented on tests/vm_test.go:
```diff
@@ -1921,6 +1922,91 @@ status:
+	Context("when node becomes unhealthy", Serial, func() {
+		var nodeName string
+
+		AfterEach(func() {
```
We can drop the `Unschedulable = false`. However, we can't drop the taint cleanup. This is due to the fact that `CleanNodes` only restore nodes with label `kubevirt.io/schedulable = true`. Nevertheless, when we set the taint `node.kubernetes.io/unschedulable=NoExecute`, the label `kubevirt.io/schedulable` is set to `false`. Therefore, the `CleanNodes` function won't revert the taint.

0xFelix replied:
Can you enhance `CleanNodes` to remove the taint as well?

jcanocan replied:
The taint has been dropped.

jcanocan commented on tests/vm_test.go:
```diff
@@ -1921,6 +1922,91 @@ status:
+
+			By("Disable scheduling in the node")
+
+			node.Spec.Unschedulable = true
```
Done.

jcanocan commented on tests/vm_test.go:
```diff
@@ -1921,6 +1922,91 @@ status:
+
+			By("The controller should recreate the launcher pod and move the VMi to phase READY")
+
+			Eventually(ThisVMI(vmi), 200*time.Second, 2*time.Second).Should(HaveConditionTrue(v1.VirtualMachineInstanceReady))
```
This is a really nice shortcut! Thanks. Done.

jcanocan commented on tests/libnode/node.go:
```diff
@@ -256,14 +267,30 @@ func Taint(nodeName string, key string, effect k8sv1.TaintEffect) {
+}
 
-	patchBytes, err := strategicpatch.CreateTwoWayMergePatch(old, newJson, node)
+func Untaint(nodeName string, key string) {
```
See: https://github.com/kubevirt/kubevirt/pull/10961#discussion_r1455649899

0xFelix commented on tests/vm_test.go:
```diff
@@ -1921,6 +1921,56 @@ status:
+
+		It("[Serial] the VMs running in that node should be respawned", func() {
+			By("Starting VM")
+			// We need to set RunningStrategy to either Always or ReRunOnFailure, the startVM function sets
+			// the RunningStrategy to Always.
```
I'd drop this comment, we want a running VM, rest doesn't matter.

jcanocan replied:
Done.

0xFelix commented on tests/vm_test.go:
```diff
@@ -1921,6 +1921,56 @@ status:
+
+			By("Delete the pod to simulate that the node has been rebooted")
+			err = kubevirt.Client().CoreV1().Pods(testsuite.GetTestNamespace(vmi)).Delete(context.Background(), pod.Name,
+				metav1.DeleteOptions{GracePeriodSeconds: pointer.Int64(0)})
```
Use `ptr.To` instead?

jcanocan replied:
Done.

0xFelix commented on tests/vm_test.go:
```diff
@@ -1921,6 +1921,56 @@ status:
+			Eventually(ThisVMI(vmi), 200*time.Second, 2*time.Second).Should(HaveConditionTrue(v1.VirtualMachineInstanceReady))
+
+			By("Check if the VMI has been recreated, this only possible if the VMI has been on FAILED Phase")
+			fetch := ThisVMI(vmi)
```
Replace by regular call to virtClient.VirtualMachines.Get? Looks unusual.

jcanocan replied:
Done.

0xFelix commented on tests/libnode/node.go:
```diff
@@ -380,3 +408,11 @@ func GetWorkerNodesWithCPUManagerEnabled(virtClient kubecli.KubevirtClient) []k8
 	return nodeList.Items
 }
+
+func SimulateUnhealthyNode(virtCli kubecli.KubevirtClient, nodeName string) {
```
I would inline this, instead of adding this func.

jcanocan replied:
The function has been dropped.

0xFelix commented on tests/vm_test.go:
```diff
@@ -1921,6 +1921,56 @@ status:
+			}, 30*time.Second, 2*time.Second).Should(BeTrue())
+
+			By("Delete the pod to simulate that the node has been rebooted")
+			err = kubevirt.Client().CoreV1().Pods(testsuite.GetTestNamespace(vmi)).Delete(context.Background(), pod.Name,
```
Use virtClient here?

jcanocan replied:
Done.

jcanocan commented:
V6:
- Changed the strategy to trigger the desired behavior in the test. Now instead of Taint the node, which is a bit aggressive and flaky prune, we just disable all virt-handler communications with the outside world and manually remove the launcher pod.
- Extracted `blackHolePod`, `getVirtHandler` and `getReadyHandlerPods` functions.

0xFelix commented on tests/utils.go:
I'm torn about adding things to `test/utils.go`. I don't think we should do it.

0xFelix commented on tests/utils.go:
```diff
@@ -2076,3 +2084,43 @@ func GetDefaultVirtHandlerDaemonSet(namespace string, config *util.KubeVirtDeplo
+	return GetDaemonSet(virtCli, componentName, namespace).Status.DesiredNumberScheduled
+}
+
+func BlackHolePod(getter GetterPodsFunc, virtCli kubecli.KubevirtClient, componentName, namespace, nodeName string, addOrDel BlackHoleAction) {
```
Can we have funcs with proper names to decide what to do instead of passing `addOrDel` strings? Can we move this func to `libpod`?

jcanocan replied:
Done.

0xFelix commented on tests/utils.go:
```diff
@@ -2076,3 +2084,43 @@ func GetDefaultVirtHandlerDaemonSet(namespace string, config *util.KubeVirtDeplo
 	return components.NewExportProxyDeployment(namespace, config.GetImageRegistry(), config.GetImagePrefix(), config.GetExportProxyVersion(), "", "", "", config.VirtExportProxyImage, config.GetImagePullPolicy(), config.GetImagePullSecrets(), config.GetVerbosity(), config.GetExtraEnv())
 }
+
+func GetDaemonSet(virtCli kubecli.KubevirtClient, componentName, namespace string) *v12.DaemonSet {
+	daemonSet, err := virtCli.AppsV1().DaemonSets(namespace).Get(context.Background(), componentName, metav1.GetOptions{})
+	ExpectWithOffset(1, err).NotTo(HaveOccurred())
+
+	return daemonSet
+}
+
+func GetReadyDaemonSetReplicas(virtCli kubecli.KubevirtClient, componentName, namespace string) func() int32 {
+	return func() int32 {
+		replicas := GetDaemonSet(virtCli, componentName, namespace).Status.NumberReady
+		return replicas
+	}
+}
+
+func GetDesiredDaemonSetReplicas(virtCli kubecli.KubevirtClient, componentName, namespace string) int32 {
+	return GetDaemonSet(virtCli, componentName, namespace).Status.DesiredNumberScheduled
+}
```
Can we inline those? They look to be relatively simple.

jcanocan replied:
I've changed the way `virt-handler` pod readiness is checked. So I've dropped these functions.

0xFelix commented on tests/virt_control_plane_test.go:
```diff
@@ -233,48 +231,31 @@ var _ = Describe("[Serial][ref_id:2717][sig-compute]KubeVirt control plane resil
-			}
-
-			blackHolePodFunc := func(addOrDel string) {
+			handlerPodGetter := func(virtCli kubecli.KubevirtClient, componentName, namespace, nodeName string) *k8sv1.PodList {
```
This adds an unnecessary closure. If we really need it, please add it as separate func to this file.

jcanocan replied:
I've better used the closures as we discussed.

0xFelix commented on tests/virt_control_plane_test.go:
```diff
@@ -233,48 +231,31 @@ var _ = Describe("[Serial][ref_id:2717][sig-compute]KubeVirt control plane resil
 
 			It("should fail health checks when connectivity is lost, and recover when connectivity is regained", func() {
-				desiredDeamonsSetCount := getVirtHandler().Status.DesiredNumberScheduled
+				desiredDeamonsSetCount := tests.GetDesiredDaemonSetReplicas(virtCli, componentName, flags.KubeVirtInstallNamespace)
```
typo

```suggestion
				desiredDaemonSetCount := tests.GetDesiredDaemonSetReplicas(virtCli, componentName, flags.KubeVirtInstallNamespace)
```

jcanocan replied:
Dropped.

0xFelix commented on tests/vm_test.go:
```diff
@@ -1921,6 +1923,72 @@ status:
+	Context("[Serial] when node becomes unhealthy", Serial, func() {
+		const componentName = "virt-handler"
+		var nodeName string
+		var numberOfVirtHandler int32
```
Might be better to initialize it here. Otherwhise it could be undefined if the test case fails before initializing it. (possibly adds flakiness)

jcanocan replied:
`numberOfVirtHandler` has been dropped. And regarding `nodeName` I can't know the value before the VMI is scheduled. I can change the VM creation to a `BeforeEach` if that's helpful.

0xFelix commented on tests/vm_test.go:
```diff
@@ -1921,6 +1923,72 @@ status:
+			nodeName = vmi.Status.NodeName
+			numberOfVirtHandler = tests.GetDesiredDaemonSetReplicas(virtClient, componentName, flags.KubeVirtInstallNamespace)
+			tests.BlackHolePod(handlerPodGetter, virtClient, componentName, flags.KubeVirtInstallNamespace, nodeName, tests.BlackHoleActionAdd)
+			Eventually(tests.GetReadyDaemonSetReplicas(virtClient, componentName, flags.KubeVirtInstallNamespace),
```
How do we know it was exactly the virt-handler on the VMI's node, that went down?

jcanocan replied:
You are right! Nice catch. I've changed how the virt-handler readiness check is performed

0xFelix commented on tests/vm_test.go:
```diff
@@ -1921,6 +1923,72 @@ status:
+			err = virtClient.CoreV1().Pods(pod.Namespace).Delete(context.Background(), pod.Name, metav1.DeleteOptions{
+				GracePeriodSeconds: ptr.To(int64(0)),
+			})
+			Expect(err).ToNot(HaveOccurred(), "Should delete virt-launcher")
```
IMO you can drop the description in this line.

jcanocan replied:
Done.

0xFelix commented on tests/vm_test.go:
```diff
@@ -1921,6 +1923,72 @@ status:
+			By("The controller should recreate the launcher pod and move the VMI to phase READY")
+			Eventually(ThisVMI(vmi), 200*time.Second, 2*time.Second).Should(HaveConditionTrue(v1.VirtualMachineInstanceReady))
+
+			By("Waiting for VMI to be running")
+			Eventually(ThisVMI(vmi), 300*time.Second, 1*time.Second).Should(BeRunning())
```
Isn't it already running, when we asserted it is Ready?

0xFelix commented on tests/vm_test.go:
```diff
@@ -1921,6 +1923,72 @@ status:
+
+			// This only possible if the VMI has been on FAILED Phase
+			By("Check if the VMI has been recreated")
+			Eventually(func() types.UID {
```
I don't think Eventually is still required here. The VMI won't change anymore. Either it was recreated or not at this point.

jcanocan replied:
In other words, it needs some time to be updated after the VMI is running.

0xFelix replied:
Ok, leaving that for others. Is it expected behavior for the VMI to change its `UID` that late? (VM is already in Ready state)

jcanocan replied:
You were right. I was my fault because I wasn't waiting until the VMI was recreated at all, only in the last Eventually statement. Done.

0xFelix commented on tests/vm_test.go:
```diff
@@ -1921,6 +1923,72 @@ status:
+			Eventually(ThisVMI(vmi), 300*time.Second, 1*time.Second).Should(BeRunning())
+
+			By("Waiting for VM to be ready")
+			Eventually(ThisVM(vm), 360*time.Second, 1*time.Second).Should(beReady())
```
How do we ensure all of this happened in a timely manner? 300 + 360 secs is a lot of time.

jcanocan replied:
I've dropped the `BeRunning()` check. But I'm not sure if we have a way to ensure that. However, this should be faster than that. Therefore, I've reduced the timeouts.

0xFelix commented on tests/vm_test.go:
```diff
@@ -1921,6 +1923,72 @@ status:
+		var nodeName string
+		var numberOfVirtHandler int32
+
+		handlerPodGetter := func(virtCli kubecli.KubevirtClient, componentName, namespace, nodeName string) *k8sv1.PodList {
```
Again, this adds an unnecessary closure.

jcanocan replied:
Same as above.

0xFelix commented on tests/libpod/render.go:
```diff
@@ -198,3 +209,32 @@ func RenderTargetcliPod(name, disksPVC string) *v1.Pod {
 	}
 }
+
+func AddServiceBackHoleForKubernetesApiToPod(pods *v1.PodList, virtCli kubecli.KubevirtClient, componentName string) {
```
```suggestion
func AddKubernetesApiBlackhole(pods *v1.PodList, virtCli kubecli.KubevirtClient, componentName string) {
```

jcanocan replied:
Done.

0xFelix commented on tests/libpod/render.go:
```diff
@@ -198,3 +209,32 @@ func RenderTargetcliPod(name, disksPVC string) *v1.Pod {
+	serviceBlackHoleForKubernetesApiToPod(pods, virtCli, componentName, BlackHoleActionAdd)
+}
+
+func DeleteServiceBlackHoleForKubernetesApiToPod(pods *v1.PodList, virtCli kubecli.KubevirtClient, componentName string) {
```
```suggestion
func DeleteKubernetesApiBlackhole(pods *v1.PodList, virtCli kubecli.KubevirtClient, componentName string) {
```

0xFelix replied:
Call it `containerName` instead of `componentName`?

jcanocan replied:
Done.

0xFelix commented on tests/libpod/render.go:
Can you add these new functions to a new file `blackhole.go` instead please?

jcanocan replied:
Done.

0xFelix commented on tests/libpod/render.go:
```diff
@@ -20,19 +20,30 @@
 	"kubevirt.io/kubevirt/tests/testsuite"
 )
 
+type BlackHoleAction string
+
+const (
+	BlackHoleActionAdd BlackHoleAction = "add"
+	BlackHoleActionDel BlackHoleAction = "del"
+)
+
```
Make all of this private resp. hide it? Or better get rid of it alltogether.

jcanocan replied:
Done.

0xFelix commented on tests/libpod/render.go:
```diff
@@ -198,3 +209,32 @@ func RenderTargetcliPod(name, disksPVC string) *v1.Pod {
+	serviceBlackHoleForKubernetesApiToPod(pods, virtCli, componentName, BlackHoleActionDel)
+}
+
+func serviceBlackHoleForKubernetesApiToPod(pods *v1.PodList, virtCli kubecli.KubevirtClient, componentName string, addOrDel BlackHoleAction) {
```
You can do `virtClient, err := kubecli.GetKubevirtClient()` so you don't have to pass the `virtClient` down the whole chain of funcs.

jcanocan replied:
Done.

0xFelix commented on tests/libpod/render.go:
```diff
@@ -198,3 +209,32 @@ func RenderTargetcliPod(name, disksPVC string) *v1.Pod {
+	serviceBlackHoleForKubernetesApiToPod(pods, virtCli, componentName, BlackHoleActionDel)
+}
+
+func serviceBlackHoleForKubernetesApiToPod(pods *v1.PodList, virtCli kubecli.KubevirtClient, componentName string, addOrDel BlackHoleAction) {
```
Instead of `addOrDel` you could make it a bool called `present`.

`present == true` --> add the blackhole
`present == false` --> remove the blackhole

jcanocan replied:
Done.

0xFelix commented on tests/libpod/render.go:
```diff
@@ -20,19 +20,30 @@
+	"context"
 	"fmt"
 
+	"github.com/onsi/gomega"
```
```suggestion
	. "github.com/onsi/gomega"
```

Import it like it is done in other files?

jcanocan replied:
Done.

0xFelix commented on tests/virt_control_plane_test.go:
```diff
@@ -245,17 +244,10 @@ var _ = Describe("[Serial][ref_id:2717][sig-compute]KubeVirt control plane resil
 			}
 
-			blackHolePodFunc := func(addOrDel string) {
+			handlerPodGetter := func() *k8sv1.PodList {
```
```suggestion
			getHandlerPods := func() *k8sv1.PodList {
```

jcanocan replied:
Done.

0xFelix commented on tests/vm_test.go:
```diff
@@ -1921,6 +1924,66 @@ status:
+		It("[Serial] the VMs running in that node should be respawned", func() {
+
+			By("Starting VM")
+			vm := startVM(virtClient, createVM(virtClient, libvmi.NewAlpine()))
```
```suggestion
			vm := startVM(virtClient, createVM(virtClient, libvmi.NewCirros()))
```

Cirros should suffice to reduce resource usage?

0xFelix commented on tests/vm_test.go:
```diff
@@ -1921,6 +1924,66 @@ status:
+			nodeName = vmi.Status.NodeName
+			oldUID := vmi.UID
+
+			By("Blocking virt-handler for updating the VMI")
```
```suggestion
			By("Blocking virt-handler from reconciling the VMI")
```

jcanocan replied:
Done.

0xFelix commented on tests/vm_test.go:
```diff
@@ -1921,6 +1924,66 @@ status:
+			pod, err := libvmi.GetPodByVirtualMachineInstance(vmi, vmi.Namespace)
+			Expect(err).ToNot(HaveOccurred())
+
+			By("Simulating lost of the virt-launcher")
```
```suggestion
			By("Simulating loss of the virt-launcher")
```

jcanocan replied:
Done.

0xFelix commented on tests/vm_test.go:
```diff
@@ -1921,6 +1924,65 @@ status:
+			Expect(err).ToNot(HaveOccurred())
+
+			By("The controller should recreate the launcher pod and move the VMI to phase READY")
+			Eventually(ThisVMI(vmi), 60*time.Second, 2*time.Second).Should(HaveConditionTrue(v1.VirtualMachineInstanceReady))
```
How long will the VMI have a condition other than Ready? E.g. going from Ready -> !Ready -> Ready, would it make sense to assert it went to something other than Ready first?

0xFelix replied:
Instead of observing the VMI we could instead wait for the VM to become not ready? And then become ready again?

jcanocan replied:
I've added a check to test if the VMI is !Ready and if the VMI is on Scheduled state at some point.

jcanocan commented:
I've changed the `AfterEach` by `DeferCleanup` to test if it works as expected. We can further discuss which cleanup strategy works better for our use case.

0xFelix commented on tests/libpod/blackhole.go:
```diff
@@ -0,0 +1,54 @@
+	. "github.com/onsi/gomega"
+
+	v1 "k8s.io/api/core/v1"
+	v12 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
nit: Usually imported as `metav1`

jcanocan replied:
Done.

0xFelix commented on tests/libpod/blackhole.go:
```diff
@@ -0,0 +1,54 @@
+	serviceBlackHoleForKubernetesApiToPod(pods, containerName, false)
+}
+
+func serviceBlackHoleForKubernetesApiToPod(pods *v1.PodList, containerName string, present bool) {
```
```suggestion
func kubernetesApiServiceBlackhole(pods *v1.PodList, containerName string, present bool) {
```

jcanocan replied:
Done.

0xFelix commented on tests/libpod/blackhole.go:
```diff
@@ -0,0 +1,54 @@
+}
+
+func serviceBlackHoleForKubernetesApiToPod(pods *v1.PodList, containerName string, present bool) {
+	var addOrDel string
```
nit: Move to L32

jcanocan replied:
Done.

0xFelix commented on tests/libpod/blackhole.go:
```diff
@@ -0,0 +1,54 @@
+}
+
+func getKubernetesApiServiceIp(virtClient kubecli.KubevirtClient) (string, error) {
+	kubernetesServiceName := "kubernetes"
+	kubernetesServiceNamespace := "default"
```
Make these const

jcanocan replied:
Done.

0xFelix commented on tests/libpod/blackhole.go:
```diff
@@ -0,0 +1,54 @@
+	kubernetesServiceNamespace := "default"
+
+	kubernetesService, err := virtClient.CoreV1().Services(kubernetesServiceNamespace).Get(context.Background(), kubernetesServiceName, v12.GetOptions{})
+	if err != nil {
+		return "", err
+	}
```
Use `Expect(err).NotTo(HaveOccurred())` since we are in test code?

jcanocan replied:
Done.

0xFelix commented on tests/libpod/render.go:
```diff
@@ -28,11 +28,18 @@ import (
 	"kubevirt.io/kubevirt/tests/testsuite"
 )
 
+/*
```
Remove this comment

jcanocan replied:
Ups, my bad. Done.

0xFelix commented on tests/vm_test.go:
```diff
@@ -1921,6 +1924,63 @@ status:
+			oldUID := vmi.UID
+
+			By("Blocking virt-handler from reconciling the VMI")
+			libpod.AddKubernetesApiBlackhole(getHandlerNodePod(nodeName), componentName)
```
nit: You could potentially save the result of `getHandlerNodePod(nodeName)` in a var.

jcanocan replied:
Done.

0xFelix commented on tests/vm_test.go:
```diff
@@ -1921,6 +1924,63 @@ status:
+			By("The controller should recreate the launcher pod and move the VMI to phase Scheduled")
+			Eventually(ThisVMI(vmi), 60*time.Second, 2*time.Second).Should(BeInPhase(v1.Scheduled))
+
+			// This only possible if the VMI has been on FAILED Phase
```
```suggestion
			// This is only possible if the VMI has been in phase Failed
```

jcanocan replied:
Done.

0xFelix commented on tests/vm_test.go:
```diff
@@ -1921,6 +1924,63 @@ status:
+			Expect(err).ToNot(HaveOccurred())
+
+			By("The VM should not be Ready")
+			Eventually(ThisVM(vm), 60*time.Second, 1*time.Second).ShouldNot(beReady())
+
+			By("The controller should recreate the launcher pod and move the VMI to phase Scheduled")
+			Eventually(ThisVMI(vmi), 60*time.Second, 2*time.Second).Should(BeInPhase(v1.Scheduled))
```
Did you try out lower values for the timeout? We want it to be fast. If this slows down significantly it could be considered a regression. Maybe add a comment about it here.

0xFelix reviewed (commented):
We're getting there!

0xFelix commented on tests/vm_test.go:
```diff
@@ -1921,6 +1924,64 @@ status:
+			By("Blocking virt-handler from reconciling the VMI")
+			virtHandlerPod := getHandlerNodePod(nodeName)
+			libpod.AddKubernetesApiBlackhole(virtHandlerPod, componentName)
+			Eventually(getHandlerNodePod(nodeName).Items[0], 120*time.Second, time.Second, HaveConditionFalse(k8sv1.PodReady))
```
Forget about that, we need the current state of the Pod.

jcanocan commented on tests/libpod/render.go:
```diff
@@ -28,7 +28,6 @@ import (
 	v13 "kubevirt.io/api/core/v1"
 
 	"kubevirt.io/kubevirt/pkg/pointer"
-
```
Done.

jcanocan commented on tests/vm_test.go:
```diff
@@ -1921,6 +1924,64 @@ status:
+		}
+
+		It("[Serial] the VMs running in that node should be respawned", func() {
+
```
Done.

jcanocan commented on tests/vm_test.go:
```diff
@@ -1921,6 +1924,64 @@ status:
+	Context("[Serial] when node becomes unhealthy", Serial, func() {
+		const componentName = "virt-handler"
+
+		getHandlerNodePod := func(nodeName string) *k8sv1.PodList {
```
Done.

0xFelix reviewed (approved):
Thanks!

/lgtm

@xpivarc Want to give this a look again?

jcanocan commented:
/retest-required

davidvossel commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -729,15 +730,17 @@ func (c *VMIController) updateStatus(vmi *virtv1.VirtualMachineInstance, pod *k8
 	// VMI is owned by virt-handler, so patch instead of update
 	if vmi.IsRunning() || vmi.IsScheduled() {
-		patchBytes, err := prepareVMIPatch(vmi, vmiCopy)
+		patchBytes, err := prepareVMIPatch(vmi, vmiCopy, pod)
```
it is more accurate to pass in `vmiPodExists` here rather than `pod`.

jcanocan replied:
Done.

davidvossel commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -870,6 +873,17 @@ func prepareVMIPatch(oldVMI, newVMI *virtv1.VirtualMachineInstance) ([]byte, err
 		log.Log.V(3).Object(oldVMI).Infof("Patching VMI activePods")
 	}
 
+	// Apply this patch because the VMI is in an invalid state. If phase is Running or Scheduled
+	// the virt-launcher Pod needs to exist.
+	if (oldVMI.Status.Phase == virtv1.Running || oldVMI.Status.Phase == virtv1.Scheduled) &&
+		newVMI.Status.Phase == virtv1.Failed && pod == nil {
```
This makes it look like we could potentially modify the `vmi.Status.Phase` within this controller but not want to actually write that change. If that is the case, then the logic elsewhere is wrong and needs to be addressed.

xpivarc replied:
@davidvossel Yes, it is self-defensiveness. I don't think that the handover is clear enough for people to not make mistakes. What root issue would be covered? This defensiveness would essentially make the "mistake" a no-op, or not?
Do you think this is making it too complex? Previously if you would make a mistake, moving the VMI to Failed after the handoff, the patch would not be generated.

mhenriks replied:
Yeah I think I agree with @davidvossel that this seems a little awkward.  Why not always add the patch if `oldVMI.Status.Phase != newVMI.Status.Phase`

0xFelix replied:
I also have to agree with @davidvossel. This is introducing a no-op for situations, where the wrong write to the VMI's status should have been caught during review.

jcanocan replied:
I agree with this idea `oldVMI.Status.Phase != newVMI.Status.Phase`. Thanks to all for sharing your concerns about it. I've changed the if statement, as @mhenriks suggested.

xpivarc commented on tests/vm_test.go:
```diff
@@ -1921,8 +1924,67 @@ status:
+			Eventually(ThisVM(vm), 30*time.Second, 1*time.Second).ShouldNot(beReady())
+
+			By("The controller should recreate the launcher pod and move the VMI to phase Scheduled")
+			Eventually(ThisVMI(vmi), 30*time.Second, 2*time.Second).Should(BeInPhase(v1.Scheduled))
```
Can't this be flaky? or do we assume that the Pod will not move to `Running` within 2 seconds?

jcanocan replied:
If I'm not wrong, please correct otherwise, since the `Virt-handler` is not online at this point, the VMI can't transition to `Running`:

https://github.com/jcanocan/kubevirt/blob/a93aec57644a23189733f77a9e829a7fd2084e13/pkg/virt-handler/vm.go#L3307

It's true that it may happen that the VMI is rescheduled in another node and then moved to `Running` fast enough that the `Eventually` does not pick up this state. But that can be addressed with an affinity rule.

xpivarc replied:
We do not pin the VM into the specific node. Affinity seems like a hack, I think we can just `Eventually` check the `UID` as mentioned before, nothing more should be needed.

jcanocan replied:
Added `NodeSelector` rule.

jcanocan replied:
Ah okay. Sounds good to me.

jcanocan replied:
Dropped `Eventually(ThisVMI(vmi), 30*time.Second, 2*time.Second).Should(BeInPhase(v1.Scheduled))` and added `UID` `Eventually` check.

davidvossel reviewed (commented):
/approve

The controller logic looks solid to me. I'll leaning on other reviewers here to ensure the test coverage is complete as part of their final lgtm

jcanocan commented:
> /approve
> 
> The controller logic looks solid to me. I'll leaning on other reviewers here to ensure the test coverage is complete as part of their final lgtm

Thanks a lot! :blush:

xpivarc reviewed (approved):
/hold cancel

0xFelix commented:
/cherry-pick release-1.1
